### ISSUE TITLE
Customized Sift/Polyphen scores with VEP

### ISSUE THREAD

**Issue by eyeamnice** [Ensembl/ensembl-vep#772]
I downloaded cache files from ensemblgenomes which does not contain sift/polyphen scores. The info.txt files in the cache directory does not contain sift b or polyphen b. How do I generate those scores and make sure VEP can use the scores for annotation? Specifically, what are the steps to follow to customize VEP to annotate my variants with those scores?

**Comment by at7** [comment]
Hello,

we calculate SIFT scores for a selection of species listed [here](https://www.ensembl.org/info/genome/variation/prediction/protein_function.html#sift). For which species would you like to obtain SIFT scores? Maybe it would be possible to include them in a future release. If you have annotations available outside of the cache files you could make use of VEP's custom annotation option which is described [here](https://www.ensembl.org/info/genome/variation/prediction/protein_function.html#sift). 

Best wishes,
Anja

**Comment by eyeamnice** [comment]
Actually, I am working with plants. I need to calculate scores for Glycine max. I have looked at the link you provided, but nothing describes how I can use my own calculated scores. How do I ensure that VEP uses any scores I may have calculated? I also thought VEP recognizes to annotate with a score only if the info.txt file contains entries of shift b and polyphen b. If possible, could you provide me with an example command line running vep for a non-human specie that uses a user calculated sift score?

Regards,
Peter

**Comment by at7** [comment]
Dear Peter,
we run an internal pipeline to calculate SIFT scores for all possible missense variants for a set of species. The scores are stored in our databases and where available dumped into VEP cache files. I can check if there are any plans to create a variation database for Glycine max in the future.

I'm also sorry for providing the wrong link to our custom annotation documentation. The correct link is [here](https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html).
If you have your own scores it would be worth investigating if you could store them in VCF format for example. Then you can make use of VEP's custom annotation option.

[Here](https://www.ensembl.org/info/docs/tools/vep/script/vep_options.html) is an example for running VEP for non-vertebrate species. Just go to the info box: Running VEP on non-vertebrate species.

Best wishes,
Anja

**Comment by eyeamnice** [comment]
Hi Anja,
I am wondering if it was possible to put out information on how VEP calls or is able to find and use the calculated SIFT scores stored in your database. Since I am working on an organism that you do not have scores for, I have generated a sift database and calculated the score. Right now, VEP still does not use these scores. In what format are the scores stored or dumped in the cache files for VEP to recognize? 
Here is my

**Comment by eyeamnice** [comment]
I also noticed that issue 776 raised a question about creating cache files. If more people are asking such questions, maybe it will be helpful to create some documentation that will help. That way users can create their own cache files if they have all the data required especially for non-model organisms to run custom VEP

**Comment by at7** [comment]
Hi Peter,
you can read about how we store the SIFT scores [here](https://www.ensembl.org/info/genome/variation/prediction/protein_function.html#nsSNP_data_format). When we generate our VEP cache files we dump the prediction matrix for each translation into the cache files.

I think for your use case it would be worth trying to store your SIFT scores in a VCF file and then use our custom annotation from VCF file option.

We will definitely discuss how we can make our VEP cache dumping pipeline more accessible to our users. At the moment we only have our [internal pipeline](https://github.com/Ensembl/ensembl-vep/blob/release/100/modules/Bio/EnsEMBL/VEP/Pipeline/DumpVEP/DumpVEP_conf.pm) which depends on the [ensembl-hive ](https://github.com/Ensembl/ensembl-hive) job scheduling API.

Best wishes,
Anja

**Comment by at7** [comment]
Dear Peter,

I was reminded of a [VEP plugin](https://github.com/Ensembl/VEP_plugins/blob/release/100/PolyPhen_SIFT.pm) which retrieves PolyPhen and SIFT predictions from a locally constructed sqlite database. Here is a [module](https://github.com/Ensembl/ensembl-variation/blob/release/100/modules/Bio/EnsEMBL/Variation/Pipeline/ProteinFunction/RunSift.pm) from our internal pipeline which is used to populate a database with SIFT results and could point you the code you need to populate your own database. In which format are your SIFT results currently?

Best wishes,
Anja

**Comment by eyeamnice** [comment]
Hi Anja,

My sift results are in text format currently. They are tab delimited. I can also generate sift scores in vcf format. I should be okay with a code that can help me populate the database from any of the sift score format (text or vcf). Custom annotation with VEP could get tricky, without some degree of carefulness, for non-model organisms e.g certain plants. The code to populate the db and any other helpful information from you will be greatly appreciated. 

Thank you
Isaac

**Comment by at7** [comment]
Dear Peter,

did you get your SIFT scores from https://sift.bii.a-star.edu.sg/sift4g/public//Glycine_max/? If that is the case have you tried using the VCF annotation option: https://sift.bii.a-star.edu.sg/sift4g/AnnotateVariants.html? You could run VEP first and then use the output, which you can specify to be VCF, as input for the SIFT annotation step.

Best wishes,
Anja

**Comment by eyeamnice** [comment]
Thanks @at7 for that information. I think that is doable. I will give it a try. If that works, how do I generate the summary report in html format? Will that still be possible? You know when I run VEP first, it generates the html summary. By the time I run sift on the output of VEP, I will not have an updated summary report. Do you know if I can generate that in a standalone method?

**Comment by eyeamnice** [comment]
Hi Anja,
I used the output of vep as input to sift4g but it did not generate the desired result. The results from vep were not annotated in the SIFT result. At this point, I am not sure what to try next.

**Comment by at7** [comment]
I'm not familiar with sift4G. However, there are two things you could check: Is your VEP output file a VCF file? Did you run VEP with --vcf? And secondly, it could be related to an assembly version difference. The latest version in https://sift.bii.a-star.edu.sg/sift4g/public//Glycine_max/ is V1.0.28. But the latest assembly version in [Ensembl](https://plants.ensembl.org/Glycine_max/Info/Index) is already V2.1. I wonder if you can create SIFT files for the latest assembly if you haven't already or if you can use an earlier version of your [GFF file](ftp://ftp.ensemblgenomes.org/pub/release-28/plants/gff3/glycine_max/) which matches V1.0.28? Please let me know with which version of SIFT files and with which assembly version you are working.

**Comment by eyeamnice** [comment]
Hi Anja,
My VEP output and SIFT outputs are vcf files and both are now same version - V2.1. Assuming I want to use bed file for the annotation, VEP has an input format shown below:
```
chr1    10000    11000    Feature1
chr3    25000    26000    Feature2
chrX    99000    99001    Feature3
```
I tried a format with sift score results  where column 4 are the sift scores in this form:
```
chr1    10000    10000    0.01
chr1    25000    25000    0.2
chr1    99000    99000    0 
chr2    11000    11000    0.11
chr2    25000    25000    0.32
chr3    22000    22000    0.10
```
and used ```--custom mysiftscore.bed.gz,bed,,SIFT, 0`` in my vep command, the title SIFT was added to the header of the vcf output but the scores were not propagated from the bed file. Do you know the reason why?

**Comment by at7** [comment]
Have you tried --custom mysiftscore.bed.gz,bed,exact,0? You also need to be careful how you provide the coordinates in your bed file. The start is 0 based and the end is not included. For the first row in your example bed file it would mean: 1 9999 10000. And then you can use the custom option: --custom mysiftscore.bed.gz,SIFT,bed,exact,0 and in your output you should see SIFT=0.01.

However, a better option is to use custom annotation from a VCF file. In a VCF file you can also specify the base change and link it to the score as provided by the sift file coming from the sift4g tool.

For example here is my example for human:
My custom annotation file: sift_scores.vcf.gz
##fileformat=VCFv4.1
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO
1       230710048       .       A       G       .       .       SIFT=0.5

Run VEP:
./vep --input_file rs699.vcf --output_file rs699.out  \
--fasta Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz \
--gtf Homo_sapiens.GRCh38.100.chr.gtf.gz \
--custom sift_scores.vcf.gz,sift,vcf,exact,0,SIFT

My output looks like:
rs699 1:230710048 G ENSG00000135744 ENST00000366667 Transcript  missense_variant  843 803 268 M/T aTg/aCg - IMPACT=MODERATE;STRAND=-1;SOURCE=Homo_sapiens.GRCh38.100.chr.gtf.gz;sift=1:230710048-230710048;sift_FILTER=.;sift_SIFT=0.5

**Comment by eyeamnice** [comment]
I wrote a custom script to integrate SIFT scores into VEP result. I will mark this as resolved based on the output. I wish I could still generate a html summary report after adding the SIFT score. I may have to write another custom script for that.

**Comment by at7** [comment]
I'm happy to hear that you successfully added the SIFT scores to your VEP results. We don't have any plans to customise the html summary report script for including scores that have been added after VEP was run. I did check for any available tools that could help you with this and found [vcfstats](https://pypi.org/project/vcfstats/). I haven't worked with vcfstats before but from reading the documentation I think that it might be able to help you to generate the summary report for your SIFT score annotations.
